pkg/service/endpoints_controller.go

```diff
@@ -26,135 +27,280 @@ import (
-		glog.Errorf("Failed to list services: %v", err)
-		return err
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
 	}
-	var resultErr error
-	for i := range services.Items {
-		service := &services.Items[i]
 
-		if service.Spec.Selector == nil {
-			// services without a selector receive no endpoints from this controller;
-			// these services will receive the endpoints that are created out-of-band via the REST API.
-			continue
+	e.queue.Dirty(key)
+}
+
+func (e *EndpointController) worker() {
+	for {
+		key, quit := e.queue.BeginProcessing()
+		if quit {
+			return
 		}
+		e.syncService(key.(string))
+		e.queue.FinishProcessing(key)
+	}
+}
 
-		glog.V(5).Infof("About to update endpoints for service %s/%s", service.Namespace, service.Name)
-		pods, err := e.client.Pods(service.Namespace).List(labels.Set(service.Spec.Selector).AsSelector())
+func (e *EndpointController) syncService(key string) {
+	startTime := time.Now()
+	defer func() {
+		glog.V(4).Infof("Finished syncing service %q endpoints. (%v)", key, time.Now().Sub(startTime))
+	}()
+	obj, exists, err := e.serviceStore.Store.GetByKey(key)
+	if err != nil || !exists {
+		// Delete the corresponding endpoint, as the service has been deleted.
+		// TODO: Please note that this will delete an endpoint when a
+		// service is deleted. However, if we're down at the time when
+		// the service is deleted, we will miss that deletion, so this
+		// doesn't completely solve the problem. See #6877.
+		namespace, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			glog.Errorf("Need to delete endpoint with key %q, but couldn't understand the key: %v", key, err)
+			return
+		}
+		err = e.client.Endpoints(namespace).Delete(name)
 		if err != nil {
-			glog.Errorf("Error syncing service: %s/%s, skipping", service.Namespace, service.Name)
-			resultErr = err
-			continue
+			glog.Errorf("Error deleting endpoint %q: %v", key, err)
 		}
+		return
+	}
 
-		subsets := []api.EndpointSubset{}
-		for i := range pods.Items {
-			pod := &pods.Items[i]
-
-			for i := range service.Spec.Ports {
-				servicePort := &service.Spec.Ports[i]
-
-				// TODO: Once v1beta1 and v1beta2 are EOL'ed,
-				// this can safely assume that TargetPort is
-				// populated, and findPort() can be removed.
-				_ = v1beta1.Dependency
-				_ = v1beta2.Dependency
-
-				portName := servicePort.Name
-				portProto := servicePort.Protocol
-				portNum, err := findPort(pod, servicePort)
-				if err != nil {
-					glog.Errorf("Failed to find port for service %s/%s: %v", service.Namespace, service.Name, err)
-					continue
-				}
-				if len(pod.Status.PodIP) == 0 {
-					glog.Errorf("Failed to find an IP for pod %s/%s", pod.Namespace, pod.Name)
-					continue
-				}
+	service := obj.(*api.Service)
+	if service.Spec.Selector == nil {
+		// services without a selector receive no endpoints from this controller;
+		// these services will receive the endpoints that are created out-of-band via the REST API.
```

Comment:
Is it possible to update selector, such that it's empty? If yes, than I think we have a race:
1. Endpoints controller should delete endpoints (right?)
2. User might have created new endpoints already

I think this is not something we should worry about too much, but maybe we should add a comment that it's possible. WDYT?